Improve docstrings language in Metadata API

ivanayov opened the pull request:
This change updates some obvious and unnecessary fields docs in the
Metadata API with more despriptive details, as well as unifies
syntax between different classes and methods

Addresses #1600

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [n/a] Tests have been added for the bug fix or new feature
- [n/a] Docs have been added for the bug fix or new feature

ivanayov commented on tuf/api/metadata.py:
```diff
@@ -966,7 +987,11 @@ class Timestamp(Signed):
+            not break serialization from and deserialisation into a Metadta
+            object and thus aligns with the TUF specification implication that
+            unrecognized attribute-value fields should be supported
         snapshot_meta: Meta information for snapshot metadata.
```
This is one I wondered whether to edit or not. It relates to obvious and unnecessary information, but on the other side it's simple and correct. I wondered what's your opinion about explaining why it's available in the API - do you find such clarification good or rather unnecessary?

jku replied:
(reply to another comment) 
Yeah documenting meta in a comprehensive manner requires a lot of specification context so maybe we leave it as is... The MetaFile type does clearly document its content after all.

jku commented on tuf/api/metadata.py:
```diff
@@ -118,14 +118,14 @@ def __init__(self, signed: T, signatures: Dict[str, Signature]):
     @classmethod
     def from_dict(cls, metadata: Dict[str, Any]) -> "Metadata[T]":
-        """Creates Metadata object from its dict representation.
+        """Creates ``Metadata`` object from its dict representation.
```
We say "dict representation" a lot as if it's obvious what that means :)

I don't have a great alternative suggestion, just commenting.

lukpueh replied:
Would it be better non-abbreviated, i.e. "dictionary representation"? I think I used dict to keep it on one line.
Or would "json/dict representation" help people better associate it with the format described in the spec?

ivanayov replied:
I think "json/dict representation" is good. "dict" vs. "dictionary" is quite similar, while "json representation" should be obvious for everyone.

jku commented on tuf/api/metadata.py:
```diff
@@ -118,14 +118,14 @@ def __init__(self, signed: T, signatures: Dict[str, Signature]):
 
         Raises:
             KeyError: The metadata dict format is invalid.
-            ValueError: The metadata has an unrecognized signed._type field.
+            ValueError: The metadata has an unrecognized ``signed._type`` field.
```
ValueError docstring is likely wrong: there are many more ways to trigger it than wrong type
I think both errors should be documented with same vague message like "Invalid metadata argument".

jku replied:
(this btw is why I'm a bit hesitant about committing to documenting ValueError/KeyError in a way that _looks_ comprehensive: the documentation is going to be either incomplete or vague. It's not practical to do the massive amount of work of actually trying to list all possible cases where object construction can fail when we rely on the core libraries/data structures raising errors so much)

jku commented on tuf/api/metadata.py:
```diff
@@ -339,18 +340,18 @@ def verify_delegate(
         Raises:
-            UnsignedMetadataError: 'delegate' was not signed with required
-                threshold of keys for 'role_name'
+            UnsignedMetadataError: A ``delegate`` was not signed with required
```
First, this refers to "delegate" but should refer to "delegated_role".
 
Second, here's an example of indefinite article looking inaccurate: The error is raised because **the specific value** of argument 'delegated_role' is invalid. I think the original (without article) is correct. I think this is true for each "raises" docstring that mentions an argument.

jku commented on tuf/api/metadata.py:
```diff
@@ -401,7 +402,11 @@ class Signed(metaclass=abc.ABCMeta):
+            fields. It guarantees that input containing unknown fields will
+            not break serialization from and deserialisation into a Metadta
+            object and thus aligns with the TUF specification implication that
+            unrecognized attribute-value fields should be supported
```
* typo of "Metadata"
* Makes promises we might not be willing to keep (you can certainly put things into unrecognised_fields we won't serialize/deserialize successfully)
* quite long, especially if copied many times over: let's try to squeeze this into a couple of lines that describe the intent (but not necessarily mechanism) in a way that's useful to a developer. If a longer description is needed, let's do that in a single place (like module docstring)

I can try to improve this if it seems difficult

jku commented on tuf/api/metadata.py:
```diff
@@ -416,7 +421,7 @@ def _type(self) -> str:
         """The metadata expiry date::
 
-        # Use 'datetime' module to e.g. expire in seven days from now
+        # Use ``datetime`` module to e.g. expire in seven days from now
```
This is not useful: ordinary comments are not included in ReadTheDocs where the double backtick is useful. 

This applies to all instance where non-docstring comments are modified like this

jku commented on tuf/api/metadata.py:
```diff
@@ -87,7 +87,7 @@ class Metadata(Generic[T]):
 
-    Metadata[T] is a generic container type where T can be any one type of
+    ``Metadata[T]`` is a generic container type where T can be any one type of
     [Root, Timestamp, Snapshot, Targets]. The purpose of this is to allow
```
I suppose you could double-backtick all of these as well to show that they are actual python types/classes?

jku reviewed (commented):
I wish this was in multiple commits so we could easily decide to only take some changes now and leave others for discussion... This issue should maybe have been split by change type to make that clearer that things like consistent double-backticks are easier to find a solution to than other issues (like consistent use of article where we need to decide what we want). The changes do conflict with each other so I can see why you did them all at once.

Anyway, I'll review but let's discuss if it makes sense to split this or to continue tweaking this PR.

--- 

There's a lot of good changes here, just commenting on potential issues.

I'm not sold on the use of indefinite article everywhere. It's fine in some places but looks quite bad in others:
`threshold: A number of keys required to...` 
`length: A length of the target file in bytes`
those do not look right

Also the use of indefinite article is not consistent: there are still places that use definite article instead and the commit even adds a few new definite articles (see TargetFile methods for examples). I don't think 100% consistency is required but I don't understand how the article is chosen...

Overall, I think the indefinite article does not work in many places: In my opinion we should not standardize on it. I would maybe suggest a default choice of no article (since it makes docstrings shorter) and diverging from that if the context requires it.

ivanayov committed:
Improve docstrings language in Metadata API - quotes

This change unifies quotes to double backtick across docs in the
Metadata API in order to provide better visualisation

Signed-off-by: Ivana Atanasova <iyovcheva@vmware.com>

ivanayov committed:
Improve docstrings language in Metadata API - wording

This change unifies wording across docs in the Metadata API, like
Args vs. Arguments and same repetitive descriptions written
differently in different classes/methods

Signed-off-by: Ivana Atanasova <iyovcheva@vmware.com>

ivanayov committed:
Improve docstrings language in Metadata API - article

This change unifies as mush as the context allows and improves the
use of definite vs. indefinite vs. no article across docs in the
Metadata API. It sticks to no article in most cases for simplisity
and readability, but leaves definite article where it's strictly
necessary

Signed-off-by: Ivana Atanasova <iyovcheva@vmware.com>

ivanayov committed:
Update repetitive docstrings language in Metadata API

This change updates some obvious and unnecessary fields docs in the
Metadata API with more despriptive details

Signed-off-by: Ivana Atanasova <iyovcheva@vmware.com>

ivanayov commented:
@jku Thank you for the review! I've split to several commits, as you advice. It's a bit harder to address feedback, as it should go to a specific commit in a single-file update, but I agree it's a better history at the end.

ivanayov commented on tuf/api/metadata.py:
```diff
@@ -1379,13 +1373,13 @@ def from_data(
 
         Raises:
-            LengthOrHashMismatchError: Calculated length or hashes do not
+            LengthOrHashMismatchError: The calculated length or hashes do not
```
This one is probably left by mistake. Will revert it at the end while addressing other comments to save time.

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -120,18 +121,14 @@ def __init__(self, signed: T, signatures: Dict[str, Signature]):
+        Args:
             metadata: TUF metadata in dict representation.
 
-        Raises:
-            KeyError: The metadata dict format is invalid.
-            ValueError: The metadata has an unrecognized signed._type field.
-
```
@jku I think we agreed we want to document `KeyError` and `ValueError` from `from_dict()` and the constructors?
That's why I created #1819.

MVrachev replied:
@ivanayov @jku has agreed we want to document those exceptions.
I will fix that when proposing a pr for #1819.
Could you remove that change?

ivanayov replied:
@MVrachev yep, I'll revert then.

ivanayov commented:
Python setup failed on windows.

jku commented on tuf/api/metadata.py:
```diff
@@ -1140,7 +1150,7 @@ def to_dict(self) -> Dict[str, Any]:
     @staticmethod
     def _is_target_in_pathpattern(targetpath: str, pathpattern: str) -> bool:
-        """Determines whether "targetname" matches the "pathpattern"."""
+        """Determines whether ``targetname`` matches the ``pathpattern``."""
```
```suggestion
        """Determines whether ``targetpath`` matches the ``pathpattern``."""
```

jku commented on tuf/api/metadata.py:
```diff
@@ -788,14 +794,14 @@ def add_key(self, role: str, key: Key) -> None:
-            role: The name of the role, for which a signing key is removed.
-            key: The identifier of the key to be removed for 'role'.
+            role: Name of the role, for which a signing key is removed.
+            key: Identifier of the key to be removed for ``role``.
```
```suggestion
            keyid: Identifier of the key to be removed for ``role``.
```

jku commented on tuf/api/metadata.py:
```diff
@@ -1476,16 +1489,16 @@ def add_key(self, role: str, key: Key) -> None:
-            role: The name of the role, for which a signing key is removed.
-            key: The identifier of the key to be removed for 'role'.
+            role: Name of the role, for which a signing key is removed.
+            key: Identifier of the key to be removed for ``role``.
```
```suggestion
            keyid: Identifier of the key to be removed for ``role``.
```

jku reviewed (changes requested):
thanks for the nice commits: I was expecting to suggest moving some commits to a later PR... but I think I actually agree with all of these, they seem to be clear improvements. Only thing that is a required fix is a rebase (merge is not currently clean)

There are things to improve in the docs but this is a big PR already and there's no need to fix everything at once, just leaving some comments: 
* I noticed some docstring errors and left comments: none of them seem to be your fault so feel free to either fix here or leave to next PR
* The unrecognised_fields docstring feels about the right amount of detail to me -- maybe we 'd like to explain general concept in one place somewhere but individual docstrings look ok
* Many arguments still explain "what" the argument is, not "why" it is: as an example we document signer argument with `securesystemslib.signer.Signer implementation` which does not explain why signer is an argument (what is it used for).

... but all of these can be handled in future improvements. I've re-run the tests after setup-python fixed itself: they look good.

lukpueh commented:
Thanks for working on this, @ivanayov and for reviewing, @jku! Let me know if you want more eyes.

jku commented on tuf/api/metadata.py:
```diff
@@ -788,14 +794,14 @@ def add_key(self, role: str, key: Key) -> None:
 
         Raises:
-            ValueError: If 'role' doesn't exist or if 'role' doesn't include
+            ValueError: If ``role' doesn't exist or if ``role`` doesn't include
```
Oh found an error:
```suggestion
            ValueError: If ``role`` doesn't exist or if ``role`` doesn't include
```

ivanayov commented:
Thank you @jku ! I addressed your comments here, as they are related to the changes anyway. 
Changed 
```
signer: ``securesystemslib.signer.Signer`` implementation
```
 to 
```
signer: ``securesystemslib.signer.Signer`` implementation containing a
signature and a keyid of the key that signature was generated with.
```
The other changes are just the fixes that you proposed + moving some of the existing changes  to the right commit (I was a bit lazy to do that before addressing the comments 😁).

There are still some 
```
data: Metadata content.
...
filename: Path to write the file to.
etc.
```
but I think there's enough context around them and adding explanations could make the docs too verbose and might rather confuse the user by creating expectations of some deeper meaning behind. Do you think that those should be better explained? @lukpueh do you have some preference, too?

jku reviewed (changes requested):
Thank you. I think I would like to leave the signer argument improvements for later: It is significantly more difficult to explain "why" than "what" -- and this PR is long enough.

In practice:
```
signer: ``securesystemslib.signer.Signer`` implementation containing a
signature and a keyid of the key that signature was generated with.
```

this is slightly incorrect (signer does not contain a signature, it produces signatures) and still doesn't explain why there is a signer argument. Could you just leave the original description in and then we can continue tweaking these later?

lukpueh commented:
> There are still some
> 
> ```
> data: Metadata content.
> ...
> filename: Path to write the file to.
> etc.
> ```
> 
> but I think there's enough context around them and adding explanations could make the docs too verbose and might rather confuse the user by creating expectations of some deeper meaning behind. Do you think that those should be better explained? @lukpueh do you have some preference, too?

I think both are okay. Especially the latter is actually adds useful information not provided by the arg name alone.

ivanayov commented:
@jku I removed the change.

jku commented:
Thanks!
